src/Umbraco.Core/Composing/CollectionBuilderBase.cs

```diff
@@ -94,30 +95,30 @@ private void RegisterTypes(IRegister register)
         /// Creates the collection items.
         /// </summary>
         /// <returns>The collection items.</returns>
-        protected virtual IEnumerable<TItem> CreateItems(IFactory factory)
+        protected virtual IEnumerable<TItem> CreateItems(IServiceProvider serviceProvider)
         {
             if (_registeredTypes == null)
                 throw new InvalidOperationException("Cannot create items before the collection builder has been registered.");
 
             return _registeredTypes // respect order
-                .Select(x => CreateItem(factory, x))
+                .Select(x => CreateItem(serviceProvider, x))
                 .ToArray(); // safe
         }
 
         /// <summary>
         /// Creates a collection item.
         /// </summary>
-        protected virtual TItem CreateItem(IFactory factory, Type itemType)
-            => (TItem) factory.GetInstance(itemType);
+        protected virtual TItem CreateItem(IServiceProvider serviceProvider, Type itemType)
+            => (TItem) serviceProvider.GetRequiredService(itemType);
 
         /// <summary>
         /// Creates a collection.
         /// </summary>
         /// <returns>A collection.</returns>
         /// <remarks>Creates a new collection each time it is invoked.</remarks>
-        public virtual TCollection CreateCollection(IFactory factory)
+        public virtual TCollection CreateCollection(IServiceProvider serviceProvider)
         {
-            return factory.CreateInstance<TCollection>(  CreateItems(factory));
+            return ActivatorUtilities.CreateInstance<TCollection>(serviceProvider, CreateItems(serviceProvider));
```

Comment:
I just scratched my head and gave up when I saw the dicts, are the mappers supposed to share a ConcurrentDictionary or have one per each? 

Regardless of the answer to the above, it's a bit too generic of a resolve, perhaps needs a wrapper to make things more obvious, an IMapperThingerDictionary

Reply:
> This takes in a ConcurrentDictionary as a ctor argument but it cannot resolve that. TBH I actually don't know how this currently works with the MapperCollectionBuilder since each mapper there has this ConcurrentDictionary passed in.

Do you know if the dict should be shared among all mappers or one per each, perhaps a wrapper around the dependency would be good, it's a bit strange to resolve a generic collection from the container.

Reply:
Yeah the Mappers with the ConcurrentDictionary param is strange. I 100% agree that it should have an actual object/interface injected instead that perhaps just contains the concurrent dictionary instance, at least it would make it clear as to what is being injected. I had a look into v8 and the way this is currently being injected is here: https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Core/Persistence/Mappers/MapperCollectionBuilder.cs#L29 where it manually creates the instance and passes in `_maps`. So yes it is the same instance passed in to all of them.